### DISCUSSION TITLE
Top-level async function returns immediately

### DISCUSSION THREAD

**Discussion by Soundman32** [ClearFoundry/ClearScript#602] [Q&A]
I'm trying to run a JS script that has async methods, so I have a top-level async method that runs everything (see below).  
```
(async () => {
    ... <long running process>
   _bag.Success = true; // Signal that the script has completed successfully
})();

_bag.Completed = true;

```

__bag is a C# PropertyBag that is added via AddHostObject_

`engine.Execute` returns immediately and I can see `_bag.Completed` is set, but the script is still executing in the background (other C# methods and logging are seen AFTER the execute has returned).

How do I make `Execute` wait until the top-level method has completed?

EDIT:  I've added this to the C# code, but it feels wrong
```csharp
engine.Execute(script);
while (!bag.ContainsKey("Success"))
{
  await Task.Delay(100);
}
`

**Comment by ClearScriptLib** [comment]
Hi @Soundman32,

>How do I make `Execute` wait until the top-level method has completed?

When you call an async function, the return value is a [promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise). If you opt into ClearScript's task-promise interop, you can simply `await` the async operation:

```csharp
// assuming engine constructed with V8ScriptEngineFlags.EnableTaskPromiseConversion
await (Task)engine.Evaluate(@"
    (async () => {
        ... <long running process>
    })();
");
```

>I've added this to the C# code, but it feels wrong

If you prefer to use the property bag as a signaling mechanism, you can do something like this to avoid polling:

```csharp
var success = new TaskCompletionSource<object>();
bag.PropertyChanged += (_, eventArgs) => {
    if (eventArgs.PropertyName == "Success") {
        success.SetResult(bag["Success"]);
    }
};
engine.Evaluate(@"
    (async () => {
        ... <long running process>
       _bag.Success = true; // Signal that the script has completed successfully
    })();
");
Console.WriteLine(await success.Task);
```

Good luck!
